File: packages/java/anip-core/src/main/java/dev/anip/core/Composition.java
package dev.anip.core;

import java.util.List;
import java.util.Map;

/** Declarative composition for kind=composed capabilities. v0.23. See SPEC.md §4.6. */
public class Composition {
    private final String authorityBoundary;
    private final List<CompositionStep> steps;
    private final Map<String, Map<String, String>> inputMapping;
    private final Map<String, String> outputMapping;
    private String emptyResultPolicy;
    private Map<String, Object> emptyResultOutput;
    private final FailurePolicy failurePolicy;
    private final AuditPolicy auditPolicy;

    public Composition(String authorityBoundary,
                       List<CompositionStep> steps,
                       Map<String, Map<String, String>> inputMapping,
                       Map<String, String> outputMapping,
                       FailurePolicy failurePolicy,
                       AuditPolicy auditPolicy) {
        this.authorityBoundary = authorityBoundary;
        this.steps = steps;
        this.inputMapping = inputMapping;
        this.outputMapping = outputMapping;
        this.failurePolicy = failurePolicy;
        this.auditPolicy = auditPolicy;
    }

    public String getAuthorityBoundary() {
        return authorityBoundary;
    }

    public List<CompositionStep> getSteps() {
        return steps;
    }

    public Map<String, Map<String, String>> getInputMapping() {
        return inputMapping;
    }

    public Map<String, String> getOutputMapping() {
        return outputMapping;
    }

    public String getEmptyResultPolicy() {
        return emptyResultPolicy;
    }

    public Composition setEmptyResultPolicy(String emptyResultPolicy) {
        this.emptyResultPolicy = emptyResultPolicy;
        return this;
    }

    public Map<String, Object> getEmptyResultOutput() {
        return emptyResultOutput;
    }

    public Composition setEmptyResultOutput(Map<String, Object> emptyResultOutput) {
        this.emptyResultOutput = emptyResultOutput;
        return this;
    }

    public FailurePolicy getFailurePolicy() {
        return failurePolicy;
    }

    public AuditPolicy getAuditPolicy() {
        return auditPolicy;
    }
}

File: packages/csharp/src/Anip.Core/Composition.cs
using System.Text.Json.Serialization;

namespace Anip.Core;

/// <summary>One ordered step in a composed capability. v0.23. See SPEC.md §4.6.</summary>
public class CompositionStep
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";

    [JsonPropertyName("capability")]
    public string Capability { get; set; } = "";

    [JsonPropertyName("empty_result_source")]
    public bool EmptyResultSource { get; set; } = false;

    [JsonPropertyName("empty_result_path")]
    public string? EmptyResultPath { get; set; }
}

/// <summary>Per-child-outcome failure handling for composed capabilities. v0.23. See SPEC.md §4.6.</summary>
public class FailurePolicy
{
    [JsonPropertyName("child_clarification")]
    public string ChildClarification { get; set; } = "propagate";

    [JsonPropertyName("child_denial")]
    public string ChildDenial { get; set; } = "propagate";

    [JsonPropertyName("child_approval_required")]
    public string ChildApprovalRequired { get; set; } = "propagate";

    [JsonPropertyName("child_error")]
    public string ChildError { get; set; } = "fail_parent";
}

/// <summary>Audit behavior for composed capabilities. v0.23. See SPEC.md §4.6.</summary>
public class AuditPolicy
{
    [JsonPropertyName("record_child_invocations")]
    public bool RecordChildInvocations { get; set; }

    [JsonPropertyName("parent_task_lineage")]
    public bool ParentTaskLineage { get; set; }
}

/// <summary>Declarative composition for kind=composed capabilities. v0.23. See SPEC.md §4.6.</summary>
public class Composition
{
    [JsonPropertyName("authority_boundary")]
    public string AuthorityBoundary { get; set; } = "same_service";

    [JsonPropertyName("steps")]
    public List<CompositionStep> Steps { get; set; } = new();

    [JsonPropertyName("input_mapping")]
    public Dictionary<string, Dictionary<string, string>> InputMapping { get; set; } = new();

    [JsonPropertyName("output_mapping")]
    public Dictionary<string, string> OutputMapping { get; set; } = new();

    [JsonPropertyName("empty_result_policy")]
    public string? EmptyResultPolicy { get; set; }

    [JsonPropertyName("empty_result_output")]
    public Dictionary<string, object>? EmptyResultOutput { get; set; }

    [JsonPropertyName("failure_policy")]
    public FailurePolicy FailurePolicy { get; set; } = new();

    [JsonPropertyName("audit_policy")]
    public AuditPolicy AuditPolicy { get; set; } = new();
}
